File: biddo_client/lib/core/controllers/account.dart
import 'package:biddo/core/controllers/image_picker.dart';
import 'package:get/get.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';

import '../models/account.dart';
import '../models/filter.dart';
import '../repositories/account.dart';
import '../repositories/filter.dart';
import 'location.dart';

class AccountController extends GetxController with StateMixin<dynamic> {
  final account = Account().obs;
  final accountsRepository = Get.find<AccountsRepository>();
  final imagePickerController = Get.find<ImagePickerController>();
  final filtersRepository = Get.find<FiltersRepository>();
  final locationController = Get.find<LocationController>();

  RxInt accountAuctionsCount = 0.obs;
  RxInt activeAuctionsCount = 0.obs;
  RxInt closedAuctionsCount = 0.obs;

  RxInt accountBidsCount = 0.obs;
  RxInt acceptedBidsCount = 0.obs;
  RxInt rejectedBidsCount = 0.obs;

  @override
  void onClose() {
    account.value = Account();
    super.onClose();
  }

  setAccount(Account? account) {
    if (account == null) {
      this.account.value = Account();
      return;
    }
    this.account.value = account;
  }

  updateCoins(int coins) {
    account.value.coins = coins;
    account.refresh();
  }

  Future<bool> loadAccountData(bool purchasesAvailable) async {
    var loadedAccount = await accountsRepository.loadLoggedIn();
    if (loadedAccount == null) {
      return false;
    }

    account.value = loadedAccount;
    loadAccountStats();

    // if (purchasesAvailable) {
    //   await Purchases.logIn(account.value.id);
    // }

    if (loadedAccount.locationLatLng != null &&
        loadedAccount.locationPretty != null) {
      locationController.setMarkerLatLong(loadedAccount.locationLatLng);
      locationController.setLocation(loadedAccount.locationPretty ?? '');
    }

    return true;
  }

  Future<bool> askForVerification() async {
    var askedSuccessful = await accountsRepository.askForVerification();
    if (!askedSuccessful) {
      return askedSuccessful;
    }

    account.value.verificationRequestedAt = DateTime.now();
    account.refresh();

    return askedSuccessful;
  }

  Future<void> loadAccountStats() async {
    var accountStats = await accountsRepository.loadStats();
    if (accountStats == null) {
      return;
    }

    updateAccountAuctionsCountFromStats(accountStats);
  }

  Future<void> updateAccountAllowedNotifications(
      AccountNotifications notifications) {
    account.value.allowedNotifications = notifications;
    return accountsRepository.update(account.value);
  }

  Future<bool> deleteAccount() async {
    return await accountsRepository.delete(account.value);
  }

  Future<bool> updateAccountCurrency(String currencyId) async {
    account.value.selectedCurrencyId = currencyId;
    account.refresh();
    var result = await accountsRepository.update(account.value);
    return result != null;
  }

  Future<bool> saveFCMToken(String token) async {
    account.value.deviceFCMToken = token;
    var result = await accountsRepository.update(account.value);
    return result != null;
  }

  Future<void> updateAccountMeta(AccountMetadata meta) {
    account.value.meta = meta;
    account.refresh();
    return accountsRepository.update(account.value);
  }

  Future<void> saveLocationToAccount(LatLng? location, String? prettyLocation) {
    account.value.locationLatLng = location;
    account.value.locationPretty = prettyLocation;
    account.refresh();
    return accountsRepository.update(account.value);
  }

  Future<bool> updateNameAndProfilePicture(String newName) async {
    var galleryAsset = imagePickerController.galleryAssets.isNotEmpty
        ? (await imagePickerController.getGalleryAssetPaths()).first
        : null;

    var cameraAsset = imagePickerController.cameraAssets.isNotEmpty
        ? imagePickerController.cameraAssets.first
        : null;

    account.value.name = newName;
    var updatedAccount = await accountsRepository.update(
      account.value,
      galleryAsset,
      cameraAsset,
    );

    if (updatedAccount != null) {
      account.value.name = newName;
      account.value.picture = updatedAccount.picture;
      account.refresh();
    }
    return updatedAccount != null;
  }

  void togglePreferredCategory(String categoryId) {
    if (categoryIsPreferred(categoryId)) {
      account.value.preferredCategoriesIds.remove(categoryId);
    } else {
      account.value.preferredCategoriesIds.add(categoryId);
    }

    account.refresh();
  }

  bool categoryIsPreferred(String categoryId) {
    return account.value.preferredCategoriesIds.contains(categoryId);
  }

  Future<bool> finishCategoriesSetup() async {
    account.value.categoriesSetupDone = true;
    account.refresh();

    var result = await accountsRepository.update(account.value);
    return result != null;
  }

  Future<Account?> loadAccountDetailsById(String accountId) async {
    return await accountsRepository.loadAccountDetails(accountId);
  }

  Future<bool> finishIntro() async {
    account.value.introDone = true;
    var result = await accountsRepository.update(account.value);
    return result != null;
  }

  Future<bool> skipIntro() async {
    account.value.introSkipped = true;
    var result = await accountsRepository.update(account.value);
    return result != null;
  }

  Future<bool> saveFilterForAccount(FilterItem filterItem) async {
    account.value.filters ??= [];
    account.value.filters!.add(filterItem);
    account.refresh();

    return await filtersRepository.create(filterItem);
  }

  Future<bool> deleteFilterForAccount(String filterId) async {
    var deleted = await filtersRepository.delete(filterId);
    if (deleted == false) {
      return deleted;
    }

    account.value.filters?.removeWhere((element) => element.id == filterId);
    account.refresh();
    return deleted;
  }

  // 7424 1190 0170 95536

  Future<bool> acceptTerms(
    bool acceptedTerms,
  ) async {
    account.value.acceptedTermsAndCondition = acceptedTerms;
    var result = await accountsRepository.update(account.value);
    return result != null;
  }

  bool isFollowingAccount(String accountId) {
    return account.value.followingAccountsIds!.contains(accountId);
  }

  Future<bool> blockAccount(String accountId) async {
    try {
      account.value.blockedAccounts ??= [];
      account.value.blockedAccounts!.add(accountId);
      account.refresh();

      await accountsRepository.blockAccount(accountId);
      return true;
    } catch (error) {
      return false;
    }
  }

  Future<bool> unblockAccount(String accountId) async {
    try {
      account.value.blockedAccounts?.remove(accountId);
      account.refresh();

      await accountsRepository.unblockAccount(accountId);
      return true;
    } catch (error) {
      return false;
    }
  }

  void clearAccount() {
    account.value = Account();
  }

  void updateAccountAuctionsCountFromStats(dynamic accountStats) {
    accountAuctionsCount.value = accountStats['auctions'];
    activeAuctionsCount.value = accountStats['activeAuctions'];
    closedAuctionsCount.value = accountStats['closedAuctions'];
    accountBidsCount.value = accountStats['bids'];
    acceptedBidsCount.value = accountStats['acceptedBids'];
    rejectedBidsCount.value = accountStats['rejectedBids'];
  }
}

File: core/controllers/account.ts
import { AccountsRepository } from '../repositories/account'
import { AccountNotifications } from '../domain/account'
import { AppStore } from '../store'
import { compressFile } from '@/utils/compressor'
import { Category } from '../domain/category'
import { AuthController } from './auth'
import { AuctionsController } from './auctions'
import { runInAction } from 'mobx'

class AccountController {
  async finishCategoriesSetup(categories: Category[]) {
    try {
      const account = AppStore.accountData!
      runInAction(() => {
        account.categoriesSetupDone = true
        account.preferredCategoriesIds = categories.map((c) => c.id)
      })

      const updatedAccount = await AccountsRepository.update(account)
      if (!updatedAccount) {
        return false
      }

      AuctionsController.refreshRecommendations()
      return true
    } catch (error) {
      console.error(`Error finishing categories setup: ${error}`)
      return false
    }
  }

  async reloadAuthUser() {
    const userDetails = await AccountsRepository.loadLoggedIn()
    if (!userDetails) {
      return
    }

    AppStore.setAccountDetails(userDetails)
  }

  async askForVerification() {
    const asked = await AccountsRepository.askForVerification()
    if (!asked) {
      return false
    }

    const account = AppStore.accountData!
    runInAction(() => {
      account.verificationRequestedAt = new Date()
    })

    return true
  }

  async saveLocationToAccount(lat?: number, lng?: number, locationPretty?: string) {
    try {
      const account = AppStore.accountData
      if (!account) {
        return false
      }

      if (account.locationLatLng?.lat === lat && account.locationLatLng?.lng === lng) {
        return true
      }

      runInAction(() => {
        account.locationLatLng = lat && lng ? { lat, lng } : undefined
        account.locationPretty = locationPretty ?? ''
      })

      const updatedAccount = await AccountsRepository.update(account)
      if (!updatedAccount) {
        return false
      }

      return true
    } catch (error) {
      console.error(`Error saving location to account: ${error}`)
      return false
    }
  }

  async updateAccountCurrency(currencyId: string) {
    try {
      const account = AppStore.accountData!
      runInAction(() => {
        account.selectedCurrencyId = currencyId
      })

      await AccountsRepository.update(account)
      return true
    } catch (error) {
      console.error(`Error updating account currency: ${error}`)
      return false
    }
  }

  async finishInto() {
    try {
      const account = AppStore.accountData!
      runInAction(() => {
        account.introDone = true
      })
      await AccountsRepository.update(account)
      return true
    } catch (error) {
      console.error(`Error finishing intro: ${error}`)
      return false
    }
  }

  async skipIntro() {
    try {
      const account = AppStore.accountData!
      runInAction(() => {
        account.introSkipped = true
      })
      await AccountsRepository.update(account)
      return true
    } catch (error) {
      console.error(`Error skipping intro: ${error}`)
      return false
    }
  }

  async acceptTerms(acceptedTerms: boolean) {
    try {
      const account = AppStore.accountData!
      if (!account) {
        return
      }
      runInAction(() => {
        account.acceptedTermsAndCondition = acceptedTerms
      })

      await AccountsRepository.update(account)
      return true
    } catch (error) {
      console.error(`Error accepting terms: ${error}`)
      return false
    }
  }

  async loadAccountStats() {
    try {
      const account = AppStore.accountData!
      if (!account) {
        return
      }

      const stats = await AccountsRepository.loadStats()
      if (stats) {
        this.updateAccountAuctionsCountFromStats(stats)
      }
    } catch (error) {
      console.error(`Error loading account stats: ${error}`)
    }
  }

  async updateAccountAllowedNotifications(notifications: AccountNotifications) {
    try {
      if (!AppStore.accountData?.id) {
        return
      }

      runInAction(() => {
        AppStore.accountData!.allowedNotifications = notifications
      })

      await AccountsRepository.update({ ...AppStore.accountData })
      return true
    } catch (error) {
      console.error(`Error updating account notifications: ${error}`)
      return false
    }
  }

  async blockAccount(accountId: string) {
    try {
      if (!AppStore.accountData?.id) {
        return
      }

      await AccountsRepository.blockAccount(accountId)
      runInAction(() => {
        AppStore.accountData!.blockedAccounts = AppStore.accountData!.blockedAccounts ?? []
        AppStore.accountData!.blockedAccounts.push(accountId)
      })

      return true
    } catch (error) {
      console.error(`Error blocking account: ${error}`)
      return false
    }
  }

  async unblockAccount(accountId: string) {
    try {
      if (!AppStore.accountData?.id) {
        return
      }

      await AccountsRepository.unblockAccount(accountId)
      runInAction(() => {
        AppStore.accountData!.blockedAccounts = AppStore.accountData!.blockedAccounts?.filter(
          (id) => id !== accountId
        )
      })

      return true
    } catch (error) {
      console.error(`Error unblocking account: ${error}`)
      return false
    }
  }

  async updateNameAndProfilePicture(name?: string, asset?: File) {
    if (!name && !asset) {
      return false
    }

    let compressedAsset: File | null = null
    if (asset) {
      compressedAsset = await compressFile(asset)
    }

    try {
      const account = AppStore.accountData
      if (!account) {
        return false
      }

      if (name) {
        runInAction(() => {
          account.name = name
        })
      }

      const updatedAccount = await AccountsRepository.update(account, compressedAsset ?? undefined)
      if (!updatedAccount) {
        return false
      }
      if (asset) {
        runInAction(() => {
          AppStore.accountData!.picture = updatedAccount?.picture
        })
      }
      return true
    } catch (error) {
      console.error(`Error updating name and profile picture: ${error}`)
      return false
    }
  }

  async deleteAccount() {
    try {
      const deleted = await AccountsRepository.delete()
      if (!deleted) {
        return false
      }
      await AuthController.logout()
      return true
    } catch (error) {
      console.error(`Failed to delete account: ${error}`)
      return false
    }
  }

  updateAccountAuctionsCountFromStats(accountStats: Record<string, number>) {
    runInAction(() => {
      AppStore.accountStats.allAuctionsCount = accountStats.auctions
      AppStore.accountStats.activeAuctions = accountStats.activeAuctions
      AppStore.accountStats.closedAuctions = accountStats.closedAuctions
      AppStore.accountStats.allBidsCount = accountStats.bids
      AppStore.accountStats.acceptedBids = accountStats.acceptedBids
      AppStore.accountStats.rejectedBids = accountStats.rejectedBids
    })
  }
}

const accountController = new AccountController()
export { accountController as AccountController }
